### ISSUE TITLE
Weird mismatched assets bug when importing entries with related assets 

### ISSUE THREAD

**Issue by ldstevens** [craftcms/cms#1673]
This is a **wild** bug that I've been dealing with for the past few weeks on and off, and haven't been able to solve, and I'd like to mention it here to see if you guys have any pointers on getting to the bottom of it.

Here's the bug in a nutshell:
- Setup clean Craft install
- Use Feed Me *or* Sprout Import to import several hundred entries, each with a related image asset
- ~10% of the imported entries will have mismatched images, and sometimes it will be different entries with different wrong assets.

This happens...
- No matter which add-on I use (Sprout Import or Feed Me)
- No matter which image import strategy I use (rebuild indexes, import via HTTP, etc)
- No matter which setup I'm on (local dev, or Digital Ocean via Cloudways dev server). That said, Josh (@engram-design) the Feed Me creator can't reproduce it locally, but does have another client experiencing the issue.

I've tried running imports with different versions of PHP (5.6, 7) and MySQL (5.5, 5.6, MariaDB 10) on my dev machine all with the same behavior.

I'm now wondering if there's something specific to Craft (or Yii) that could be related, or if it's an obscure environmental issue.

It's something I'd like to get to the bottom of, because I'd like to migrate more sites to Craft in due course :) Any suggestions about where to look next would be much appreciated! 

You can see my write up here:
https://github.com/engram-design/FeedMe/issues/199
and here with images of mismatched assets:
https://github.com/Pennebaker/craftcms-thearchitect/issues/46 (NB it wasn't The Architect as I initially suspected here.)

Thanks!

**Comment by engram-design** [comment]
Diving into the cause for this issue, I believe I've found the cause. As @ldstevens explains, this occurs when using Feed Me to import entries, with data mapped to an Asset field. To double-check our findings, we've imported the filename of the asset into a plain text field, as well as an Asset field. This allows us to cross-reference why an Asset field would match an incorrect Asset with a provided filename.

After extensive testing, I've isolated the issue to `AssetsFieldType.php:prepValueFromPost()`. In this function, it creates a hash of the element this Asset field belongs to (in this case an Entry). 

```
if ( ($hash = $this->element ? spl_object_hash($this->element) : null) && isset($this->_prepValueFromPostResults[$hash]) ) {
    return $this->_prepValueFromPostResults[$hash];
}
```

The above code creates a hash of the Entry, and uses this hash as a key to a private variable `_prepValueFromPostResults` which stores previous element content. This, I assume, is done for performance.

The hash it generates is something like `000000006b855451000000001d0f7e66`, and the problem is, running over a test of 100 entry imports, some hashes are identical, even if their entry content and attributes are different. Having identical hashes mean that its obviously going to grab the cached copy. For instance, the below two elements produce the same hash (I've omitted the full object for brevity).

```
Array
(
    [id] => 3880
    [slug] => root-item-446
    [sectionId] => 5
    [typeId] => 5
)
```
Produces a hash of `0000000073e34c3e000000006143f70e`.

```
Array
(
    [id] => 4760
    [slug] => root-item-4408
    [sectionId] => 5
    [typeId] => 5
```
Also produces a hash of `0000000073e34c3e000000006143f70e`.

Now, looking into the documentation of [spl_object_hash](http://php.net/manual/en/function.spl-object-hash.php) (I'm no expert), and one of the comments mentions that the hash doesn't actually look at the contents of the EntryModel.

```
Note that the contents (properties) of the object are NOT hashed by the function, merely its internal handle and handler table pointer. This is sufficient to guarantee that any two objects simultaneously co-residing in memory will have different hashes. Uniqueness is not guaranteed between objects that did not reside in memory simultaneously, for example:

var_dump(spl_object_hash(new stdClass()), spl_object_hash(new stdClass()));

Running this alone will usually generate the same hashes, since PHP reuses the internal handle for the first stdClass after it has been dereferenced and destroyed when it creates the second stdClass.
```

So it seems like this can, and does happen, when running a large import.

In summary, this can be solved temporarily by replacing the above code with `$hash = $this->element ? spl_object_hash($this->element) : null;`, but thats hardly a solution.

Do you have any thoughts regarding this? Do you believe the issue lies with Feed Me, or that using `spl_object_hash` is not reliable? Interested to hear your opinion!

**Comment by brandonkelly** [comment]
Nice find, @engram-design! Just fixed this for the next release.

**Comment by engram-design** [comment]
@brandonkelly Great to hear! That certainly was a curly issue to track down. Cheers!
